File: app/android/app/src/main/kotlin/org/wordbridge/wordbridge/ClipAudio.kt
package org.wordbridge.wordbridge

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioTrack
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.concurrent.thread

/**
 * Plays a buffer this app synthesised, now.
 *
 * The neural voice produces samples rather than calling a platform
 * synthesiser, so something has to play them, and it has to start immediately:
 * every tap on the board comes through here.
 *
 * Newest wins, matching every other speech path in the app.
 */
class ClipAudio(messenger: io.flutter.plugin.common.BinaryMessenger) {
  private val channel = MethodChannel(messenger, "org.wordbridge/clip_audio")
  private val main = Handler(Looper.getMainLooper())

  private var track: AudioTrack? = null

  /** Answered when the clip finishes, is replaced, or is stopped. */
  private var pending: MethodChannel.Result? = null

  init {
    channel.setMethodCallHandler { call, result -> handle(call, result) }
  }

  private fun handle(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "play" -> {
        val wav = call.argument<ByteArray>("wav")
        val volume = call.argument<Double>("volume") ?: 1.0
        if (wav == null) {
          result.error("arguments", "play needs wav bytes", null)
          return
        }
        play(wav, volume.toFloat(), result)
      }
      "stop" -> {
        finish()
        result.success(null)
      }
      else -> result.notImplemented()
    }
  }

  private fun play(wav: ByteArray, volume: Float, result: MethodChannel.Result) {
    finish()

    // The Dart side sends a WAV so both platforms take the same bytes; the
    // header is 44 bytes and AudioTrack wants the samples on their own.
    val header = 44
    if (wav.size <= header) {
      result.success(null)
      return
    }
    val sampleRate =
      ByteBuffer.wrap(wav, 24, 4).order(ByteOrder.LITTLE_ENDIAN).int

    val attributes =
      AudioAttributes.Builder()
        .setUsage(AudioAttributes.USAGE_MEDIA)
        .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
        .build()
    val format =
      AudioFormat.Builder()
        .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
        .setSampleRate(sampleRate)
        .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
        .build()

    val samples = wav.size - header
    val playing =
      AudioTrack(
        attributes,
        format,
        maxOf(samples, AudioTrack.getMinBufferSize(
          sampleRate, AudioFormat.CHANNEL_OUT_MONO, AudioFormat.ENCODING_PCM_16BIT)),
        AudioTrack.MODE_STREAM,
        AudioManager.AUDIO_SESSION_ID_GENERATE,
      )
    playing.setVolume(volume)
    track = playing
    pending = result
    playing.play()

    thread(name = "wordbridge-clip") {
      try {
        playing.write(wav, header, samples)
        // Blocks until what was written has actually been heard, which is what
        // makes waiting on this the same promise as waiting on platform TTS.
        playing.write(ByteArray(0), 0, 0)
      } catch (_: IllegalStateException) {
        // Stopped from under us. The caller is answered below either way.
      }
      main.post { if (track === playing) finish() }
    }
  }

  private fun finish() {
    track?.let {
      try {
        it.pause()
        it.flush()
        it.stop()
      } catch (_: IllegalStateException) {
      }
      it.release()
    }
    track = null
    pending?.success(null)
    pending = null
  }
}

File: tools/neuralbench/ios/Runner/ClipAudio.swift
import AVFoundation
import Flutter

/// Plays a buffer this app synthesised, now.
///
/// The neural voice produces samples rather than calling into a platform
/// synthesiser, so something has to play them. `audioplayers` cannot: its
/// darwin side answers `setSourceBytes is not currently implemented on iOS`,
/// which leaves writing a file between a tap and a word — the cost §4.4
/// rejected this whole route for.
///
/// A held `AVAudioEngine` rather than an `AVAudioPlayer` per clip. Measured on
/// the floor device: building a player from data and preparing it costs about
/// 117 ms before any sound, on every tap, which is the kind of number the
/// cache exists to avoid. A running engine with a player node scheduled into
/// it starts in single-digit milliseconds because everything expensive has
/// already happened.
///
/// Newest wins, matching every other speech path in the app: a user tapping
/// quickly wants the word under their finger, not a queue of the ones they
/// have moved on from.
class ClipAudio: NSObject {
  private let engine = AVAudioEngine()
  private let node = AVAudioPlayerNode()
  private var running = false

  /// Completes when the clip finishes, is replaced, or is stopped. Held so a
  /// superseded clip's caller is answered rather than left waiting for audio
  /// that will never arrive.
  private var pending: FlutterResult?

  /// Which clip is the current one. A buffer's completion handler fires on an
  /// audio thread after the node has been stopped, so it has to be able to
  /// tell "I finished" from "something newer replaced me".
  private var generation: Int = 0

  static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "org.wordbridge/clip_audio",
      binaryMessenger: registrar.messenger())
    let instance = ClipAudio()
    channel.setMethodCallHandler { call, result in
      instance.handle(call, result: result)
    }
  }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "play":
      guard let arguments = call.arguments as? [String: Any],
        let pcm = arguments["pcm"] as? FlutterStandardTypedData,
        let sampleRate = arguments["sampleRate"] as? Int,
        let volume = arguments["volume"] as? Double
      else {
        result(FlutterError(code: "arguments", message: "play needs pcm bytes", details: nil))
        return
      }
      play(pcm.data, sampleRate: Double(sampleRate), volume: Float(volume), result: result)

    case "stop":
      finish(answering: true)
      result(nil)

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Brings the engine up once, at the format the model produces.
  ///
  /// The session itself is already configured for playback with the ringer
  /// switch ignored — flutter_tts does that at startup, and an AAC device that
  /// goes silent because somebody flicked a switch is a user who cannot speak.
  private func start(sampleRate: Double) throws {
    let format = AVAudioFormat(
      standardFormatWithSampleRate: sampleRate, channels: 1)!
    if running, engine.isRunning,
      node.outputFormat(forBus: 0).sampleRate == sampleRate {
      return
    }
    if running {
      engine.stop()
      engine.disconnectNodeOutput(node)
    } else {
      engine.attach(node)
    }
    engine.connect(node, to: engine.mainMixerNode, format: format)
    engine.prepare()
    try engine.start()
    running = true
  }

  private func play(
    _ pcm: Data, sampleRate: Double, volume: Float, result: @escaping FlutterResult
  ) {
    finish(answering: true)

    let frames = pcm.count / 2
    guard frames > 0 else {
      result(nil)
      return
    }

    do {
      try start(sampleRate: sampleRate)
    } catch {
      result(FlutterError(code: "play", message: "\(error)", details: nil))
      return
    }

    let format = AVAudioFormat(
      standardFormatWithSampleRate: sampleRate, channels: 1)!
    guard
      let buffer = AVAudioPCMBuffer(
        pcmFormat: format, frameCapacity: AVAudioFrameCount(frames))
    else {
      result(nil)
      return
    }
    buffer.frameLength = AVAudioFrameCount(frames)

    // Signed 16-bit little-endian to the float format the engine mixes in.
    pcm.withUnsafeBytes { (raw: UnsafeRawBufferPointer) in
      let samples = raw.bindMemory(to: Int16.self)
      let out = buffer.floatChannelData![0]
      for i in 0..<frames {
        out[i] = Float(Int16(littleEndian: samples[i])) / 32768.0
      }
    }

    generation += 1
    let mine = generation
    pending = result
    node.volume = volume

    node.scheduleBuffer(buffer, at: nil, options: []) { [weak self] in
      DispatchQueue.main.async {
        guard let self = self, self.generation == mine else { return }
        self.answer()
      }
    }
    node.play()
  }

  /// Stops whatever is playing and answers whoever was waiting on it.
  private func finish(answering: Bool) {
    generation += 1
    if running {
      node.stop()
    }
    if answering { answer() }
  }

  private func answer() {
    guard let pending = pending else { return }
    self.pending = nil
    pending(nil)
  }
}
